### PR TITLE
PT-1924: Use rouding policies

### PR DESCRIPTION
**Pull request by krankenbro**

### CODE REVIEW

File: src/VirtoCommerce.CartModule.Data/Services/DefaultShoppingCartTotalsCalculator.cs [hunk 1]
```diff
@@ -11,6 +13,14 @@ namespace VirtoCommerce.CartModule.Data.Services
     /// </summary>
     public class DefaultShoppingCartTotalsCalculator : IShoppingCartTotalsCalculator
     {
+        private readonly ICurrencyService _currencyService;
+        private ISettingsManager _settingsManager;
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CartModule.Data/Services/DefaultShoppingCartTotalsCalculator.cs:17]:
```suggestion
        private readonlyISettingsManager _settingsManager;
```

**Review by Vectorfield4 — APPROVED** [review]

File: src/VirtoCommerce.CartModule.Data/Services/DefaultShoppingCartTotalsCalculator.cs [hunk 2]
```diff
@@ -103,30 +113,34 @@ public virtual void CalculateTotals(ShoppingCart cart)
...
+                cart.ShippingSubTotalWithTax = currency.RoundingPolicy.RoundMoney(cart.ShippingSubTotalWithTax, currency);
+                cart.PaymentTotal = currency.RoundingPolicy.RoundMoney(cart.PaymentTotal, currency);
+                cart.PaymentTotalWithTax = currency.RoundingPolicy.RoundMoney(cart.PaymentTotalWithTax, currency);
+                cart.PaymentSubTotal = currency.RoundingPolicy.RoundMoney(cart.PaymentSubTotal, currency);
+                cart.PaymentSubTotalWithTax = currency.RoundingPolicy.RoundMoney(cart.PaymentSubTotalWithTax, currency);
+                cart.PaymentDiscountTotal = currency.RoundingPolicy.RoundMoney(cart.PaymentDiscountTotal, currency);
+                cart.PaymentDiscountTotalWithTax = currency.RoundingPolicy.RoundMoney(cart.PaymentDiscountTotalWithTax, currency);
 
-            cart.Total = cart.SubTotal + cart.ShippingSubTotal + cart.TaxTotal + cart.PaymentSubTotal + cart.FeeTotal - cart.DiscountTotal;
+                cart.Total = cart.SubTotal + cart.ShippingSubTotal + cart.TaxTotal + cart.PaymentSubTotal + cart.FeeTotal - cart.DiscountTotal;
```
Comment (eMazeika) [inline src/VirtoCommerce.CartModule.Data/Services/DefaultShoppingCartTotalsCalculator.cs:142]: Why not setting total without any condition?

File: tests/VirtoCommerce.CartModule.Tests/UnitTests/CartTotalsCalculationTests.cs [hunk 3]
```diff
@@ -11,117 +11,117 @@ namespace VirtoCommerce.CartModule.Test.UnitTests
...
-            Assert.Equal(49.86m, cart.PaymentSubTotalWithTax);
-            Assert.Equal(150.01m, cart.TaxTotal);
-            Assert.Equal(176.47m, cart.DiscountTotal);
-            Assert.Equal(197.65m, cart.DiscountTotalWithTax);
-            Assert.Equal(13.64m, cart.FeeTotal);
-            Assert.Equal(15.28m, cart.FeeTotalWithTax);
-            Assert.Equal(14.68m, cart.FeeWithTax);
-            Assert.Equal(1413.18m, cart.Total);
-        }
+        //[Fact]
```
Comment (eMazeika) [inline tests/VirtoCommerce.CartModule.Tests/UnitTests/CartTotalsCalculationTests.cs:14]: Can delete or fix the tests?

File: src/VirtoCommerce.CartModule.Data/Services/DefaultShoppingCartTotalsCalculator.cs [hunk 4]
```diff
@@ -103,30 +113,34 @@ public virtual void CalculateTotals(ShoppingCart cart)
...
-            cart.PaymentTotal = Math.Round(cart.PaymentTotal, 2, MidpointRounding.AwayFromZero);
-            cart.PaymentTotalWithTax = Math.Round(cart.PaymentTotalWithTax, 2, MidpointRounding.AwayFromZero);
-            cart.PaymentSubTotal = Math.Round(cart.PaymentSubTotal, 2, MidpointRounding.AwayFromZero);
-            cart.PaymentSubTotalWithTax = Math.Round(cart.PaymentSubTotalWithTax, 2, MidpointRounding.AwayFromZero);
-            cart.PaymentDiscountTotal = Math.Round(cart.PaymentDiscountTotal, 2, MidpointRounding.AwayFromZero);
-            cart.PaymentDiscountTotalWithTax = Math.Round(cart.PaymentDiscountTotalWithTax, 2, MidpointRounding.AwayFromZero);
+            if (_settingsManager.GetValue<bool>(StoreModule.Core.ModuleConstants.Settings.General.EnablePriceRoundingForTotalsCalculation.Name, true))
+            {
+                //Need to round all cart totals
+                var currency = _currencyService.GetAllCurrenciesAsync().GetAwaiter().GetResult().First(c => c.Name == cart.Currency);
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CartModule.Data/Services/DefaultShoppingCartTotalsCalculator.cs:119]:
```suggestion
                var currency = _currencyService.GetAllCurrenciesAsync().GetAwaiter().GetResult().First(c => c.Code == cart.Currency);
```

**Review by Vectorfield4 — CHANGES_REQUESTED** [review]

File: tests/VirtoCommerce.CartModule.Tests/UnitTests/CartTotalsCalculationTests.cs (hunk 3 above)
Comment (Vectorfield4) [reply tests/VirtoCommerce.CartModule.Tests/UnitTests/CartTotalsCalculationTests.cs:14]: https://virtocommerce.atlassian.net/browse/PT-3614

**Review by Vectorfield4 — APPROVED** [review]

**Pull request merged**